File: electron/modules/downloader/disk-manager.ts
import * as fs from "fs";
import * as path from "path";
import { getDiskInfo as nativeGetDiskInfo, getAllDisk as nativeGetAllDisk, DiskType } from "../../i10r-addon";

const GB = 1024 ** 3;

export type DriveCategory = "internal_ssd" | "external_ssd" | "hdd" | "usb" | "unknown";

export interface DiskInfo {
  path: string;
  available: number;
  total: number;
  isSSd: boolean;
  isRemovable: boolean;
}

export interface DiskEnvironmentInfo {
  environment: "ssd_save" | "hdd_ssd_tmp" | "hdd_only" | "unknown";
  saveDrive: DriveEnvInfo;
  tmpDrive: DriveEnvInfo | null;
}

export interface DriveEnvInfo {
  path: string;
  mediaType: "SSD" | "HDD" | "USB" | "unknown";
}

export class DiskManager {
  private usageTracker = new Map<string, number>();

  private ssdCache = new Map<string, boolean>();
  private drivesCache: { result: DiskInfo[]; ts: number } | null = null;
  private readonly DRIVES_CACHE_TTL = 30_000;

  // ─── Public API ─────────────────────────────────────────────────────────────

  async getDiskInfo(targetPath: string): Promise<DiskInfo> {
    const resolved = path.resolve(targetPath);
    const native = nativeGetDiskInfo(resolved);
    if (!native) throw new Error(`Cannot get disk info for: ${resolved}`);
    return {
      path: native.mountPoint,
      available: native.freeSpace,
      total: native.totalSpace,
      isSSd: native.diskType === DiskType.SSD || native.diskType === DiskType.NVME,
      isRemovable: native.isRemovable,
    };
  }

  categorizeDrive(diskInfo: DiskInfo): DriveCategory {
    const native = nativeGetDiskInfo(diskInfo.path);
    if (!native) return "unknown";
    if (native.diskType === DiskType.SSD || native.diskType === DiskType.NVME) {
      return native.isRemovable ? "external_ssd" : "internal_ssd";
    }
    if (native.diskType === DiskType.USB) return "usb";
    if (native.diskType === DiskType.HDD) return "hdd";
    return "unknown";
  }

  async getDriveCategoryForPath(targetPath: string): Promise<DriveCategory> {
    const info = await this.getDiskInfo(targetPath);
    return this.categorizeDrive(info);
  }

  async detectSSD(targetPath: string): Promise<boolean> {
    const key = this.driveKey(targetPath);
    if (this.ssdCache.has(key)) return this.ssdCache.get(key)!;
    const native = nativeGetDiskInfo(targetPath);
    if (!native) return false;
    const isSSD = native.diskType === DiskType.SSD || native.diskType === DiskType.NVME;
    this.ssdCache.set(key, isSSD);
    return isSSD;
  }

  /**
   * Find an SSD to use as tmp directory when useTmp is enabled.
   * Returns the SSD mount point with most free space, or null if none found.
   */
  async findTmpDir(requiredBytes: number): Promise<string | null> {
    const drives = nativeGetAllDisk();
    let best: { path: string; available: number } | null = null;
    for (const d of drives) {
      if (d.diskType !== DiskType.SSD && d.diskType !== DiskType.NVME) continue;
      if (d.freeSpace < requiredBytes) continue;
      if (!best || d.freeSpace > best.available) {
        best = { path: d.mountPoint, available: d.freeSpace };
      }
    }
    return best?.path ?? null;
  }

  async chooseTmpDir(
    savePath: string,
    requiredBytes: number,
    _fileSize: number,
    _preferredTmpDir?: string,
  ): Promise<string | null> {
    return this.findTmpDir(requiredBytes);
  }

  async hasEnoughSpace(
    savePath: string,
    firmwareSize: number,
    bufferBytes: number = 5 * GB,
    deleteOnRun: IPSWFile[] = [],
  ): Promise<{ ok: boolean; available: number; required: number; unknown?: boolean }> {
    const freeAfterDelete = deleteOnRun.reduce((a, b) => a + b.size, 0);
    const currentUsage = Array.from(this.usageTracker.values()).reduce((a, b) => a + b, 0);
    const required = Math.max(0, firmwareSize + currentUsage + bufferBytes - freeAfterDelete);
    try {
      const info = await this.getDiskInfo(savePath);
      return { ok: info.available >= required, available: info.available, required };
    } catch {
      return { ok: false, available: -1, required, unknown: true };
    }
  }

  reserveSpace(taskId: string, bytes: number): void { this.usageTracker.set(taskId, bytes); }
  releaseSpace(taskId: string): void { this.usageTracker.delete(taskId); }
  getTotalReserved(): number { return Array.from(this.usageTracker.values()).reduce((a, b) => a + b, 0); }

  async getEnvironmentInfo(savePath: string): Promise<DiskEnvironmentInfo> {
    const saveDir = path.resolve(savePath);
    const isSSD = await this.detectSSD(saveDir);

    const saveDrive: DriveEnvInfo = {
      path: this.driveKey(saveDir),
      mediaType: isSSD ? "SSD" : "HDD",
    };

    let environment: DiskEnvironmentInfo["environment"];
    let tmpDrive: DriveEnvInfo | null = null;

    if (isSSD) {
      environment = "ssd_save";
    } else {
      const tmpDir = await this.findTmpDir(1 * GB);
      if (tmpDir !== null) {
        environment = "hdd_ssd_tmp";
        tmpDrive = { path: this.driveKey(tmpDir), mediaType: "SSD" };
      } else {
        environment = "hdd_only";
      }
    }

    return { environment, saveDrive, tmpDrive };
  }

  // ─── Driver helpers ─────────────────────────────────────────────────────────

  driveKey(filePath: string): string {
    const resolved = path.resolve(filePath);
    if (process.platform === "win32") return path.parse(resolved).root.toUpperCase();
    const parts = resolved.split(path.sep).filter(Boolean);
    return path.sep + (parts[0] ?? "");
  }

  // ─── Private ─────────────────────────────────────────────────────────────────

  private resolveDir(resolved: string): string {
    try {
      return fs.existsSync(resolved) && fs.statSync(resolved).isDirectory()
        ? resolved : path.dirname(resolved);
    } catch { return path.dirname(resolved); }
  }
}

File: src-tauri/src/downloader/disk_manager.rs
use crate::downloader::types::{DiskEnvironmentInfo, DiskInfo, DriveEnvInfo, DlEnvironment, DlMediaType};
use std::collections::HashMap;
use std::path::{Path, PathBuf};
use std::sync::{Arc, Mutex};
use std::time::{Duration, Instant};

const GB: u64 = 1024 * 1024 * 1024;

pub trait DiskProbe: Send + Sync {
    fn disk_info(&self, dir: &Path) -> DiskInfo;
    fn all_drives(&self) -> Vec<DiskInfo>;
}

pub struct SysinfoDiskProbe;

impl DiskProbe for SysinfoDiskProbe {
    fn disk_info(&self, dir: &Path) -> DiskInfo {
        let disks = sysinfo::Disks::new_with_refreshed_list();
        let dir = dir.to_path_buf();
        let mut best: Option<&sysinfo::Disk> = None;
        let mut best_len = 0usize;
        for disk in disks.list() {
            let mount = disk.mount_point();
            if dir.starts_with(mount) {
                let len = mount.as_os_str().len();
                if len > best_len {
                    best_len = len;
                    best = Some(disk);
                }
            }
        }
        let result = match best {
            Some(d) => DiskInfo {
                path: d.mount_point().to_string_lossy().to_string(),
                available: d.available_space(),
                total: d.total_space(),
                is_ssd: matches!(d.kind(), sysinfo::DiskKind::SSD),
            },
            None => DiskInfo {
                path: dir.to_string_lossy().to_string(),
                available: 0,
                total: 0,
                is_ssd: false,
            },
        };
        log::info!(
            "SysinfoDiskProbe: queried={:?} matched={} available={} total={}",
            dir, result.path, result.available, result.total
        );
        result
    }

    fn all_drives(&self) -> Vec<DiskInfo> {
        let disks = sysinfo::Disks::new_with_refreshed_list();
        disks
            .list()
            .iter()
            .filter(|d| !d.is_removable())
            .map(|d| DiskInfo {
                path: d.mount_point().to_string_lossy().to_string(),
                available: d.available_space(),
                total: d.total_space(),
                is_ssd: matches!(d.kind(), sysinfo::DiskKind::SSD),
            })
            .collect()
    }
}

struct CacheEntry {
    result: DiskInfo,
    at: Instant,
}

pub struct DiskManager {
    probe: Arc<dyn DiskProbe>,
    usage_tracker: Mutex<HashMap<String, u64>>,
    disk_info_cache: Mutex<HashMap<String, CacheEntry>>,
    drives_cache: Mutex<Option<(Vec<DiskInfo>, Instant)>>,
}

const DISK_INFO_CACHE_TTL: Duration = Duration::from_secs(3);
const DRIVES_CACHE_TTL: Duration = Duration::from_secs(15);

impl DiskManager {
    pub fn new(probe: Box<dyn DiskProbe>) -> Self {
        Self {
            probe: Arc::from(probe),
            usage_tracker: Mutex::new(HashMap::new()),
            disk_info_cache: Mutex::new(HashMap::new()),
            drives_cache: Mutex::new(None),
        }
    }

    pub fn with_default_probe() -> Self {
        Self::new(Box::new(SysinfoDiskProbe))
    }

    async fn probe_disk_info(&self, dir: &Path) -> DiskInfo {
        let dir = dir.to_path_buf();
        let probe = self.probe.clone();
        tokio::task::spawn_blocking(move || probe.disk_info(&dir))
            .await
            .unwrap_or(DiskInfo {
                path: String::new(),
                available: 0,
                total: 0,
                is_ssd: false,
            })
    }

    async fn probe_all_drives(&self) -> Vec<DiskInfo> {
        let probe = self.probe.clone();
        tokio::task::spawn_blocking(move || probe.all_drives())
            .await
            .unwrap_or_default()
    }

    fn resolve_dir(&self, target: &Path) -> PathBuf {
        if target.is_dir() {
            target.to_path_buf()
        } else {
            target
                .parent()
                .map(|p| p.to_path_buf())
                .unwrap_or_else(|| target.to_path_buf())
        }
    }

    pub async fn disk_info(&self, target_path: &Path) -> DiskInfo {
        let dir = self.resolve_dir(target_path);
        let key = dir.to_string_lossy().to_string();

        {
            let cache = self.disk_info_cache.lock().unwrap();
            if let Some(entry) = cache.get(&key) {
                if entry.at.elapsed() < DISK_INFO_CACHE_TTL {
                    return entry.result.clone();
                }
            }
        }
        let result = self.probe_disk_info(&dir).await;
        self.disk_info_cache.lock().unwrap().insert(
            key,
            CacheEntry {
                result: result.clone(),
                at: Instant::now(),
            },
        );
        result
    }

    pub async fn detect_ssd(&self, target_path: &Path) -> bool {
        self.disk_info(target_path).await.is_ssd
    }

    async fn all_drives(&self) -> Vec<DiskInfo> {
        {
            let cache = self.drives_cache.lock().unwrap();
            if let Some((drives, at)) = cache.as_ref() {
                if at.elapsed() < DRIVES_CACHE_TTL {
                    return drives.clone();
                }
            }
        }
        let drives = self.probe_all_drives().await;
        *self.drives_cache.lock().unwrap() = Some((drives.clone(), Instant::now()));
        drives
    }

    pub async fn choose_tmp_dir(
        &self,
        save_path: &Path,
        required_bytes: u64,
        file_size: u64,
        preferred_tmp_dir: Option<&Path>,
    ) -> Option<PathBuf> {
        let buffer = GB;
        let min_space = required_bytes.max(file_size) + buffer;

        if let Some(preferred) = preferred_tmp_dir {
            if tokio::fs::create_dir_all(preferred).await.is_ok() {
                let info = self.disk_info(preferred).await;
                if info.is_ssd && info.available >= min_space {
                    return Some(preferred.to_path_buf());
                }
            }
        }

        let drives = self.all_drives().await;
        let candidates: Vec<&DiskInfo> = drives
            .iter()
            .filter(|d| d.is_ssd && d.available >= min_space)
            .collect();
        if candidates.is_empty() {
            return None;
        }

        let save_key = drive_key(save_path);
        let mut scored: Vec<(i64, &DiskInfo)> = candidates
            .iter()
            .map(|d| (self.score_drive(d, required_bytes, &save_key), *d))
            .collect();
        scored.sort_by(|a, b| b.0.cmp(&a.0));
        scored.first().map(|(_, d)| {
            let mut p = PathBuf::from(&d.path);
            p.push(".ipsw-tmp");
            p
        })
    }

    pub async fn has_enough_space(
        &self,
        save_path: &Path,
        firmware_size: u64,
        buffer_bytes: u64,
    ) -> (bool, u64, u64) {
        let current_usage: u64 = self.usage_tracker.lock().unwrap().values().sum();
        let required = firmware_size + current_usage + buffer_bytes;
        let info = self.disk_info(&self.resolve_dir(save_path)).await;
        log::info!(
            "has_enough_space: path={:?} resolved={:?} available={} total={} required={} firmware={} usage={} buffer={}",
            save_path, self.resolve_dir(save_path), info.available, info.total, required, firmware_size, current_usage, buffer_bytes
        );
        (info.available >= required, info.available, required)
    }

    pub fn reserve_space(&self, task_id: &str, bytes: u64) {
        self.usage_tracker
            .lock()
            .unwrap()
            .insert(task_id.to_string(), bytes);
    }

    pub fn release_space(&self, task_id: &str) {
        self.usage_tracker.lock().unwrap().remove(task_id);
    }

    pub async fn environment_info(&self, save_path: &Path) -> DiskEnvironmentInfo {
        let save_dir = self.resolve_dir(save_path);
        let is_ssd = self.detect_ssd(&save_dir).await;
        let save_mount = drive_key(&save_dir);
        let save_drive = DriveEnvInfo {
            path: save_mount,
            media_type: if is_ssd { DlMediaType::Ssd } else { DlMediaType::Hdd },
        };

        if is_ssd {
            return DiskEnvironmentInfo {
                environment: DlEnvironment::SsdSave,
                save_drive,
                tmp_drive: None,
            };
        }

        match self.choose_tmp_dir(save_path, GB, GB, None).await {
            Some(tmp_dir) => DiskEnvironmentInfo {
                environment: DlEnvironment::HddSsdTmp,
                save_drive,
                tmp_drive: Some(DriveEnvInfo {
                    path: drive_key(&tmp_dir),
                    media_type: DlMediaType::Ssd,
                }),
            },
            None => DiskEnvironmentInfo {
                environment: DlEnvironment::HddOnly,
                save_drive,
                tmp_drive: None,
            },
        }
    }

    fn score_drive(&self, drive: &DiskInfo, required_bytes: u64, save_drive_key: &str) -> i64 {
        let mut score: i64 = 0;
        let headroom_gb =
            (drive.available as i64 - required_bytes as i64) as f64 / GB as f64;
        score += (headroom_gb.max(0.0) * 2.0).floor().min(50.0) as i64;

        if !drive.path.to_uppercase().starts_with("C:") {
            score += 30;
        }
        if drive_key(Path::new(&drive.path)) == save_drive_key {
            score += 20;
        }
        let total_gb = drive.total as f64 / GB as f64;
        score += (total_gb / 50.0).floor().min(20.0) as i64;
        score
    }
}

pub fn drive_key(path: &Path) -> String {
    #[cfg(windows)]
    {
        use std::path::Component;
        for c in path.components() {
            if let Component::Prefix(p) = c {
                return p.as_os_str().to_string_lossy().to_uppercase();
            }
        }
        String::new()
    }
    #[cfg(not(windows))]
    {
        let mut parts = path.components();
        let _root = parts.next();
        match parts.next() {
            Some(first) => format!("/{}", first.as_os_str().to_string_lossy()),
            None => "/".to_string(),
        }
    }
}
